pkg/flowaggregator/certificate.go

```diff
@@ -0,0 +1,192 @@
+// limitations under the License.
+
+package flowaggregator
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+	"math/big"
+	"net"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+)
+
+const (
+	CAConfigMapName       = "flow-aggregator-ca"
+	CAConfigMapKey        = "ca.crt"
+	CAConfigMapNamespace  = "flow-aggregator"
+	ClientSecretNamespace = "flow-aggregator"
+	// #nosec G101: false positive triggered by variable name which includes "Secret"
+	ClientSecretName = "flow-aggregator-client-tls"
+)
+
+var (
+	validFrom = time.Now().Add(-time.Hour) // valid an hour earlier to avoid flakes due to clock skew
+	maxAge    = time.Hour * 24 * 365       // one year self-signed certs
+)
+
+func generateCACertKey() (*x509.Certificate, *rsa.PrivateKey, []byte, error) {
+	cert := &x509.Certificate{
+		SerialNumber: big.NewInt(1),
+		Subject: pkix.Name{
+			CommonName: fmt.Sprintf("flow-aggregator-ca@%d", time.Now().Unix()),
+		},
+		NotBefore:             validFrom,
+		NotAfter:              validFrom.Add(maxAge),
+		IsCA:                  true,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		BasicConstraintsValid: true,
+	}
+	// generate private key for CA
+	caKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	// generate CA certificate
+	caCert, err := x509.CreateCertificate(rand.Reader, cert, cert, &caKey.PublicKey, caKey)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	caPEM := new(bytes.Buffer)
+	pem.Encode(caPEM, &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: caCert,
+	})
+
+	return cert, caKey, caPEM.Bytes(), err
+}
+
+func generateCertKey(caCert *x509.Certificate, caKey *rsa.PrivateKey, isServer bool, flowAggregatorAddress string) ([]byte, []byte, error) {
+	var cert *x509.Certificate
+	if isServer {
+		if ip := net.ParseIP(flowAggregatorAddress); ip != nil {
+			cert = &x509.Certificate{
+				SerialNumber: big.NewInt(2),
+				Subject: pkix.Name{
+					CommonName: fmt.Sprintf("flow-aggregator-server-certificate@%d", time.Now().Unix()),
+				},
+				IPAddresses: []net.IP{ip},
+				NotBefore:   validFrom,
+				NotAfter:    validFrom.Add(maxAge),
+				ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+				KeyUsage:    x509.KeyUsageDigitalSignature,
+			}
+		} else {
+			cert = &x509.Certificate{
+				SerialNumber: big.NewInt(2),
+				Subject: pkix.Name{
+					CommonName: fmt.Sprintf("flow-aggregator-server-certificate@%d", time.Now().Unix()),
+				},
+				DNSNames:    []string{flowAggregatorAddress},
+				NotBefore:   validFrom,
+				NotAfter:    validFrom.Add(maxAge),
+				ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+				KeyUsage:    x509.KeyUsageDigitalSignature,
+			}
+		}
```

Comment:
please avoid the code duplication. It makes it hard to see the differences between the 2 cases and to maintain the code.
```go
cert = &x509.Certificate{
    SerialNumber: big.NewInt(2),
    Subject: pkix.Name{
        CommonName: fmt.Sprintf("flow-aggregator-server-certificate@%d", time.Now().Unix()),
    },
    NotBefore:   validFrom,
    NotAfter:    validFrom.Add(maxAge),
    ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
    KeyUsage:    x509.KeyUsageDigitalSignature,
}
if ip := net.ParseIP(flowAggregatorAddress); ip != nil {
    cert.IPAddresses = []net.IP{ip}
} else {
    cert.DNSNames = []string{flowAggregatorAddress}
}
```